query/src/api/rpc/exchange/exchange_channel_sender.rs

```diff
@@ -12,66 +12,292 @@
+                        sleep_future = n;
+
+                        if let Ok(recv_packet) = recv_message {
+                            if c_tx.send(recv_packet).await.is_err() {
+                                break 'fragment_loop;
+                            }
+
+                            continue 'fragment_loop;
+                        }
+
+                        // Disconnect channel, exit loop
+                        if to_request_server {
+                            ExchangeSender::send_progress_if_need(&ctx, &c_tx).await;
+                            ExchangeSender::send_precommit_if_need(&ctx, &c_tx).await;
+                        }
+
+                        let fragment_end = FragmentData::End(fragment_id);
+                        c_tx.send(DataPacket::FragmentData(fragment_end)).await.ok();
+                        break 'fragment_loop;
+                    }
+                };
+            }
+        }));
+
+        Ok(f_tx)
+    }
+
+    async fn send_progress_if_need(ctx: &Arc<QueryContext>, c_tx: &Sender<DataPacket>) {
+        let scan_progress = ctx.get_scan_progress();
+        let scan_progress_values = scan_progress.fetch();
+
+        if scan_progress_values.rows != 0 || scan_progress_values.bytes != 0 {
+            let progress = ProgressInfo::ScanProgress(scan_progress_values);
+            if c_tx.send(DataPacket::Progress(progress)).await.is_err() {
+                common_tracing::tracing::warn!(
+                    "Send scan progress values error, because flight connection is closed."
+                );
+                return;
+            }
+        }
+
+        let write_progress = ctx.get_write_progress();
+        let write_progress_values = write_progress.fetch();
+
+        if write_progress_values.rows != 0 || write_progress_values.bytes != 0 {
+            let progress = ProgressInfo::WriteProgress(write_progress_values);
+            if c_tx.send(DataPacket::Progress(progress)).await.is_err() {
+                common_tracing::tracing::warn!(
+                    "Send write progress values error, because flight connection is closed."
+                );
+                return;
+            }
+        }
+
+        let result_progress = ctx.get_result_progress();
+        let result_progress_values = result_progress.fetch();
+
+        if result_progress_values.rows != 0 || result_progress_values.bytes != 0 {
+            let progress = ProgressInfo::ResultProgress(result_progress_values);
+            if c_tx.send(DataPacket::Progress(progress)).await.is_err() {
+                common_tracing::tracing::warn!(
+                    "Send result progress values error, because flight connection is closed."
+                );
+            }
+        }
+    }
+
+    async fn send_precommit_if_need(ctx: &Arc<QueryContext>, c_tx: &Sender<DataPacket>) {
+        let precommit_blocks = ctx.consume_precommit_blocks();
+
+        if !precommit_blocks.is_empty() {
```

Comment:
no need this judge I think,

Reply:
precommit_blocks is empty if execute `SELECT`;